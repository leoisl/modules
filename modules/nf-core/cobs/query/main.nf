process COBS_QUERY {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/cobs:0.3.0--hdcf5f25_0' :
        'biocontainers/cobs:0.3.0--hdcf5f25_0'}"

    input:
    tuple val(meta),  path(query)
    tuple val(meta_index), path(index)

    output:
    tuple val(meta), path("matches.gz"), emit: matches
    path "versions.yml"                , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Note: xz is explicitly supported in this module as COBS indexes compress well under xz (see https://doi.org/10.1101/2023.04.15.536996)

    // variables setup
    def args = task.ext.args ?: ''
    def should_load_the_whole_index_into_RAM = args.contains("--load-complete")
    def index_is_gzip_compressed = index.toString().endsWith(".gz")
    def index_is_xz_compressed = index.toString().endsWith(".xz")
    def decompress_tool = ""
    def get_index_size_command = ""

    if (index_is_gzip_compressed) {
        decompress_tool = "gzip"
        get_index_size_command = "index_size=\$(gzip --list "$index" | tail -n1 | awk '{print \\\$2}')"
    } else if (index_is_xz_compressed) {
        decompress_tool = "xz"
        get_index_size_command = "index_size=\$(xz --list --robot "$index" | grep file | awk '{print \\\$5}')"
    }
    def command =
        """
            set -euo pipefail
        """

    // run COBS
    if (index_is_gzip_compressed || index_is_xz_compressed) {
        if (should_load_the_whole_index_into_RAM) {
            // streams compressed index to COBS
            command += get_index_size_command
            command += """
                            cobs \\
                                query \\
                                $args \\
                                -T $task.cpus \\
                                -i <(${decompress_tool} --decompress --stdout "${index}") \\
                                -f <(zcat $query) \\
                                --index-sizes \$index_size \\
                       """
        } else {
            // decompresses compressed index to disk and mmap it
            def decompressed_index = index.toString().replaceAll(".{g,x}z\$", "")
            command += """
                            ${decompress_tool} --decompress --stdout "${index}" > "${decompressed_index}"

                            cobs \\
                                query \\
                                $args \\
                                -T $task.cpus \\
                                -i $decompressed_index \\
                                -f <(zcat $query) \\

                            rm "${decompressed_index}"
                       """
        }
    }else {
        // index is not compressed - will be loaded completely into RAM (if --load-complete is passed) or memory mapped otherwise
        command += """
                        cobs \\
                            query \\
                            $args \\
                            -T $task.cpus \\
                            -i $index \\
                            -f <(zcat $query) \\
                   """
    }
    command +=
        """ | gzip > matches.gz

                cat <<-END_VERSIONS > versions.yml
                "${task.process}":
                    cobs: \$(cobs version 2>&1 | awk '{print \$3}')
                END_VERSIONS
        """
    command
}
