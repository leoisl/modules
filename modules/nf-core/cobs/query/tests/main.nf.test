nextflow_process {
    name "Test Process COBS_QUERY"
    script "../main.nf"
    process "COBS_QUERY"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "cobs"
    tag "cobs/query"
    tag "xz/compress"
    tag "cobs/classicconstruct"
    tag "pigz/compress"
    tag "cobs/compactconstruct"

    test("sarscov2 - classic - no compression - reads - fastq") {
        setup {
            run("COBS_CLASSICCONSTRUCT") {
                script "../../classicconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = COBS_CLASSICCONSTRUCT.out.index
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - compact - no compression - reads - fastq") {
        setup {
            run("COBS_COMPACTCONSTRUCT") {
                script "../../compactconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = COBS_COMPACTCONSTRUCT.out.index
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - classic - xz compression - reads - fastq") {
        setup {
            run("COBS_CLASSICCONSTRUCT") {
                script "../../classicconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
            run("XZ_COMPRESS") {
                script "../../../xz/compress/main.nf"
                process {
                    """
                    input[0] = COBS_CLASSICCONSTRUCT.out.index
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = XZ_COMPRESS.out.archive
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - compact - xz compression - reads - fastq") {
        setup {
            run("COBS_COMPACTCONSTRUCT") {
                script "../../compactconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
            run("XZ_COMPRESS") {
                script "../../../xz/compress/main.nf"
                process {
                    """
                    input[0] = COBS_COMPACTCONSTRUCT.out.index
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = XZ_COMPRESS.out.archive
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - classic with load complete - xz compression - reads - fastq") {
        config "./nextflow.load_complete.config"
        setup {
            run("COBS_CLASSICCONSTRUCT") {
                script "../../classicconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
            run("XZ_COMPRESS") {
                script "../../../xz/compress/main.nf"
                process {
                    """
                    input[0] = COBS_CLASSICCONSTRUCT.out.index
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = XZ_COMPRESS.out.archive
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - classic - gz compression - reads - fastq") {
        setup {
            run("COBS_CLASSICCONSTRUCT") {
                script "../../classicconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
            run("PIGZ_COMPRESS") {
                script "../../../pigz/compress/main.nf"
                process {
                    """
                    input[0] = COBS_CLASSICCONSTRUCT.out.index
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = PIGZ_COMPRESS.out.archive
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - compact - gz compression - reads - fastq") {
        setup {
            run("COBS_COMPACTCONSTRUCT") {
                script "../../compactconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
            run("PIGZ_COMPRESS") {
                script "../../../pigz/compress/main.nf"
                process {
                    """
                    input[0] = COBS_COMPACTCONSTRUCT.out.index
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = PIGZ_COMPRESS.out.archive
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - classic with load complete - gz compression - reads - fastq") {
        config "./nextflow.load_complete.config"
        setup {
            run("COBS_CLASSICCONSTRUCT") {
                script "../../classicconstruct/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test'], // meta map
                        file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
            run("PIGZ_COMPRESS") {
                script "../../../pigz/compress/main.nf"
                process {
                    """
                    input[0] = COBS_CLASSICCONSTRUCT.out.index
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true)
                ]
                input[1] = PIGZ_COMPRESS.out.archive
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("consistent_output")}
            )
        }
    }

    test("sarscov2 - classic - no compression - reads - fastq - stub") {
        options "-stub-run"
        when {
            process {
                """
                input[0] = [
                    [ id:'test'], // meta map,
                    file('reads.fastq.gz')
                ]
                input[1] = [
                    [ id:'test'], // meta map
                    file('test.index.cobs_classic')
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match()}
            )
        }
    }
}
