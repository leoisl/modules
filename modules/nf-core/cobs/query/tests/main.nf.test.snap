{
    "versions": {
        "content": [
            [
                "versions.yml:md5,bcff4ee09e1eab79e92a478301d31869"
            ]
        ],
        "timestamp": "2023-12-09T11:59:55.328031"
    },
    "matches": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "matches.gz:md5,d72afca994827384c775a8e3738f86bc"
                ]
            ]
        ],
        "timestamp": "2023-12-08T20:34:48.319928"
    }
}